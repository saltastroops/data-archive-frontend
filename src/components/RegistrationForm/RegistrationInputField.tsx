import * as React from "react";

/*
The registration input field responsible for user input.

Properties:
-----------
label
  The label which detail the user input field
name
  The name that uniquecly identifies the input field
onChange
  The fuction that is responsible for updating the input field value when changed
placeholder
  It gives an example of the required value
type
  The type of the input field e.g. text, password, etc.
value
  The value of the input field
*/

interface IRegistrationInputProps {
  label?: string;
  name?: string;
  onChange?: any;
  placeholder?: string;
  type?: string;
  value?: string;
}

const RegistrationInputField = (props: IRegistrationInputProps) => {
  const { label, name, onChange, placeholder, type, value } = props;
  return (
    <div className={"field"}>
      <label htmlFor={`${name}+${label}`} className={"label"}>
        {label}:
      </label>
      <div className={"control is-child"}>
        <input
          id={name}
          type={type}
          className={`${name} input`}
          name={name}
          placeholder={placeholder}
          value={value}
          onChange={onChange}
        />
      </div>
    </div>
  );
};

export default RegistrationInputField;
