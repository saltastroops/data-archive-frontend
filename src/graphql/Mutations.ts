import gql from "graphql-tag";

/**
 * Create a new user.
 *
 * The following arguments must be supplied.
 *
 * affiliation:
 *     The affiliation of the user, such as a university or an institute.
 * email:
 *     Email address. This will be stored in lower case.
 * familyName:
 *     The family name (surname).
 * givenName:
 *     The given name (first name).
 * password:
 *     The password, which must have at least 7 characters.
 * username:
 *     The username, which must not contain upper case letters.
 */
export const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $familyName: String!
    $givenName: String!
    $username: String!
    $email: String!
    $affiliation: String!
    $password: String!
  ) {
    signup(
      familyName: $familyName
      givenName: $givenName
      username: $username
      email: $email
      affiliation: $affiliation
      password: $password
    ) {
      id
    }
  }
`;

/**
 * Log the user in.
 *
 * The following arguments must be supplied
 *
 * password:
 *     The password.
 * username:
 *     The uusername.
 */
export const LOGIN_MUTATION = gql`
  mutation LOGIN_MUTATION($username: String!, $password: String!) {
    login(password: $password, username: $username) @client
  }
`;

/**
 * Log the user out.
 */
export const LOGOUT_MUTATION = gql`
  mutation LOGOUT_MUTATION {
    logout @client
  }
`;

/**
 * Update user.
 *
 * The following arguments may be supplied.
 *
 * affiliation:
 *     The affiliation of the user, such as a university or an institute.
 * email:
 *     Email address. This will be stored in lower case.
 * familyName:
 *     The family name (surname).
 * givenName:
 *     The given name (first name).
 * id:
 *     The user ID different from the currently logged in user.
 * newPassword:
 *     The new user password, which must have at lease 7 characters.
 * password:
 *     The password, which must have at least 7 characters and is required.
 * username:
 *     The username, which must not contain upper case letters.
 */
export const UPDATE_USER_MUTATION = gql`
  mutation UPDATE_USER_MUTATION(
    $id: ID
    $familyName: String
    $givenName: String
    $username: String
    $email: String
    $affiliation: String
    $password: String!
    $newPassword: String
  ) {
    updateUser(
      id: $id
      familyName: $familyName
      givenName: $givenName
      username: $username
      email: $email
      affiliation: $affiliation
      password: $password
      newPassword: $newPassword
    ) {
      id
    }
  }
`;
