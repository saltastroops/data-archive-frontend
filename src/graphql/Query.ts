import gql from "graphql-tag";

// TODO UPDATE
/**
 * Retrieving the user information.
 */
const USER_QUERY = gql`
  query USER_QUERY {
    user {
      familyName
      givenName
      roles
    }
  }
`;

export { USER_QUERY };
