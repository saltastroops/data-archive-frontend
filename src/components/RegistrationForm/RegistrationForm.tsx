import PropTypes from "prop-types";
import * as React from "react";
import { Mutation } from "react-apollo";
import styled from "styled-components";
import { SIGNUP_MUTATION } from "../../graphql/Mutations";
import { validateRegistrationField } from "../../util/RegistrationFormValidation";
import RegistrationInputField from "./RegistrationInputField";

/*
The registration form responsible for creating the new user data acrchive account.

Properties:
-----------
userInput : {
  affiliation
    The affiliation of the user, such as a university or an institute.
  confirmPassword
    A matching password that must match with the original password.
  email
    Email address. This will be stored in lower case.
  familyName
    The family name (surname).
  givenName
    The given name (first name).
  password
    The password, which must have at least 7 characters.
  username
    The username, which must not contain upper case letters.
}

errors : []
  An array of error messages if any.
*/

interface IRegistrationFormState {
  errors: string[];
  userInput: {
    affiliation: string;
    confirmPassword: string;
    email: string;
    familyName: string;
    givenName: string;
    password: string;
    username: string;
  };
}

const RegistrationFormParent = styled.form.attrs({
  className: "column is-4 is-offset-4"
})`
  && {
    padding: 1px;
  }
`;

const Heading = styled.h1.attrs({
  className: "title is-3"
})`
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
`;

const ErrorMessage = styled.p.attrs({
  className: "error tile"
})`
  && {
    text-align: left;
    margin: 3px 0 3px 0;
    padding: 2px 0 2px 0;
    background-color: hsl(348, 100%, 61%);
    color: white;
  }
`;

class RegistrationForm extends React.Component<IRegistrationFormState> {
  // Setting context prop type for accessing router field
  static contextTypes = {
    router: PropTypes.object
  };

  public state = {
    errors: [],
    userInput: {
      affiliation: "",
      confirmPassword: "",
      email: "",
      familyName: "",
      givenName: "",
      password: "",
      username: ""
    }
  };

  onHandleSubmit = async (e: any, signup: any) => {
    e.preventDefault();

    // Validates the user input fields
    const errors = validateRegistrationField(this.state.userInput);

    // Update the errors propetry of the state
    this.setState({ errors });

    // Check if there is an error, if there is abort signing up.
    if (errors.length) {
      return;
    }

    try {
      // If all fields are validated, sign up
      const res = await signup();
      // Reset the state when registering succeeded
      this.setState({
        userInput: {
          affiliation: "",
          confirmPassword: "",
          email: "",
          familyName: "",
          givenName: "",
          password: "",
          username: ""
        }
      });

      alert("Successfully registered, Login using your username and password");
      this.context.router.history.push("/login");
    } catch (error) {
      return;
    }
  };

  onHandleInputChange = (e: any) => {
    const name = e.target.name;
    const value = e.target.value;
    // Updating the userInput property of the state when input field value updates
    this.setState({
      userInput: {
        ...this.state.userInput,
        [name]: value
      }
    });
  };

  render() {
    const { errors } = this.state;
    const {
      affiliation,
      confirmPassword,
      email,
      familyName,
      givenName,
      password,
      username
    } = this.state.userInput;

    return (
      <Mutation mutation={SIGNUP_MUTATION} variables={this.state.userInput}>
        {(signup, { loading, error }) => {
          return (
            <RegistrationFormParent
              onSubmit={e => this.onHandleSubmit(e, signup)}
            >
              <Heading>Create your account</Heading>
              {error ? <ErrorMessage>{error.message}</ErrorMessage> : null}

              {errors.map(err => (
                <ErrorMessage key={err}>Error: {err}</ErrorMessage>
              ))}

              <RegistrationInputField
                name={"givenName"}
                label={"Given name (first name)"}
                onChange={this.onHandleInputChange}
                type={"text"}
                value={givenName}
              />
              <RegistrationInputField
                name={"familyName"}
                label={"Family name (surname)"}
                onChange={this.onHandleInputChange}
                type={"text"}
                value={familyName}
              />
              <RegistrationInputField
                name={"email"}
                label={"Email address"}
                onChange={this.onHandleInputChange}
                type={"text"}
                value={email}
              />
              <RegistrationInputField
                name={"username"}
                label={"Username"}
                onChange={this.onHandleInputChange}
                type={"text"}
                value={username}
              />
              <RegistrationInputField
                name={"affiliation"}
                label={"Affiliation"}
                placeholder={"E.g. University of Cape Town"}
                onChange={this.onHandleInputChange}
                type={"text"}
                value={affiliation}
              />
              <RegistrationInputField
                name={"password"}
                label={"Password"}
                placeholder={"At least 7 characters"}
                onChange={this.onHandleInputChange}
                type={"password"}
                value={password}
              />
              <RegistrationInputField
                name={"confirmPassword"}
                label={"Re-enter password "}
                onChange={this.onHandleInputChange}
                type={"password"}
                value={confirmPassword}
              />
              <button
                className="signUp button is-success is-fullwidth is-rounded"
                disabled={loading}
                type={"submit"}
              >
                {loading ? "Signing up..." : "Sign up"}
              </button>
            </RegistrationFormParent>
          );
        }}
      </Mutation>
    );
  }
}

export default RegistrationForm;
