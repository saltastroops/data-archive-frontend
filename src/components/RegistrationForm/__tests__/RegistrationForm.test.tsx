import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { SIGNUP_MUTATION } from "../../../graphql/Mutations";
import RegistrationForm from "../RegistrationForm";

// Helper function for simulating input field value change.
function inputTyping(wrapper: any, name: string, value: string) {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
}

// Initial userInput state
const initialState = {
  errors: [],
  userInput: {
    affiliation: "",
    confirmPassword: "",
    email: "",
    familyName: "",
    givenName: "",
    password: "",
    username: ""
  }
};

// Updated userInput state
const updatedState = {
  errors: [],
  userInput: {
    affiliation: "UCT",
    confirmPassword: "securepassword",
    email: "valid@email.address",
    familyName: "Smith",
    givenName: "John",
    password: "securepassword",
    username: "sj"
  }
};

// sign up mock mutation
const mocks = [
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        ...updatedState.userInput
      }
    },
    result: {
      data: {
        signup: {
          __typename: "User",
          ...updatedState.userInput,
          id: "1"
        }
      }
    }
  }
];

describe("RegistrationForm Component", () => {
  it("renders the RegistrationForm having unpopulated props with no errors", async () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={initialState.userInput}
          errors={initialState.errors}
        />
      </MockedProvider>
    );
    // Expect the snapshot to match the RegistrationForm component.
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("displays no errors if submitted inputs are all valid", async () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the affiliation input field value changes.
    inputTyping(wrapper, "affiliation", "UCT");
    // Simulate state change when the confirmPassword input field value changes.
    inputTyping(wrapper, "confirmPassword", "securepassword");
    // Simulate state change when the emailAddress input field value changes.
    inputTyping(wrapper, "email", "valid@email.address");
    // Simulate state change when the familyName input field value changes.
    inputTyping(wrapper, "familyName", "Smith");
    // Simulate state change when the givenName input field value changes.
    inputTyping(wrapper, "givenName", "John");
    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "securepassword");
    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "sj");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(7);

    // Expect the property affiliation of the state to have been updated with the correct value.
    expect(instance.state.userInput.affiliation).toBe("UCT");

    // Expect the property confirmPassword of the state to have been updated with the correct value.
    expect(instance.state.userInput.confirmPassword).toBe("securepassword");

    // Expect the property emailAddress of the state to have been updated with the correct value.
    expect(instance.state.userInput.email).toBe("valid@email.address");

    // Expect the property familyName of the state to have been updated with the correct value.
    expect(instance.state.userInput.familyName).toBe("Smith");

    // Expect the property givenName of the state to have been updated with the correct value.
    expect(instance.state.userInput.givenName).toBe("John");

    // Expect the property password of the state to have been updated with the correct value.
    expect(instance.state.userInput.password).toBe("securepassword");

    // Expect the property username of the state to have been updated with the correct value.
    expect(instance.state.userInput.username).toBe("sj");

    // Expect the button to not be clicked
    expect(wrapper.find(".signUp").text()).toContain("Sign up");

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect the button to hev beeen clicked
    expect(wrapper.find(".signUp").text()).toContain("Signing up");

    // Expect no error message.
    expect(wrapper.find(".error").length).toBe(0);
  });

  it("displays error message if submitted ivalid given name", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the givenName input field value changes.
    inputTyping(wrapper, "givenName", "");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect an error message.
    expect(wrapper.find(".error").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find(".error").text()).toContain("Given name");
    expect(wrapper.find(".error").text()).toContain("empty");
  });

  it("displays error message if submitted invalid family name", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the familyName input field value changes.
    inputTyping(wrapper, "familyName", "");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect an error message.
    expect(wrapper.find(".error").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find(".error").text()).toContain("Family name");
    expect(wrapper.find(".error").text()).toContain("empty");
  });

  it("displays error message if submitted invalid family name", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the emailAddress input field value changes.
    inputTyping(wrapper, "email", "invalid$email&address");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect an error message.
    expect(wrapper.find(".error").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find(".error").text()).toContain("Email address");
    expect(wrapper.find(".error").text()).toContain("invalid");
  });

  it("displays error message if submitted invalid username", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "sJ");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect an error message.
    expect(wrapper.find(".error").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find(".error").text()).toContain("Username");
    expect(wrapper.find(".error").text()).toContain("lowercase");
  });

  it("displays error message if submitted invalid password", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "short");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect an error message.
    expect(wrapper.find(".error").length).toBe(2);

    // Expect meaningful error message
    expect(
      wrapper
        .find(".error")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find(".error")
        .first()
        .text()
    ).toContain("7 characters");
  });

  it("displays error message if submitted invalid confirmPassword", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm
          userInput={updatedState.userInput}
          errors={updatedState.errors}
        />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "securepassword");
    // Simulate state change when the confirmPassword input field value changes.
    inputTyping(wrapper, "confirmPassword", "notmatching");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(2);

    // Simulate the submiting of the form.
    wrapper.find(".signUp").simulate("submit");

    // Expect an error message.
    expect(wrapper.find(".error").length).toBe(1);

    // Expect meaningful error message
    expect(
      wrapper
        .find(".error")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find(".error")
        .first()
        .text()
    ).toContain("do not");
  });
});
