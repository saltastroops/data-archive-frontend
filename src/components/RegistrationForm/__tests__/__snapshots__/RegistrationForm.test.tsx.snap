// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm Component renders the RegistrationForm having unpopulated props with no errors 1`] = `
<MockedProvider
  addTypename={true}
>
  <ApolloProvider
    client={
      ApolloClient {
        "__operations_cache__": Map {},
        "cache": InMemoryCache {
          "addTypename": true,
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "resultCaching": true,
          },
          "data": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "maybeBroadcastWatch": [Function],
          "optimisticData": DepTrackingCache {
            "data": Object {},
            "depend": [Function],
          },
          "silenceBroadcast": false,
          "storeReader": StoreReader {
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "executeSelectionSet": [Function],
            "executeStoreQuery": [Function],
          },
          "storeWriter": StoreWriter {},
          "typenameDocumentCache": Map {},
          "watches": Set {},
        },
        "defaultOptions": Object {},
        "disableNetworkFetches": false,
        "link": ApolloLink {
          "request": [Function],
        },
        "mutate": [Function],
        "query": [Function],
        "queryDeduplication": true,
        "reFetchObservableQueries": [Function],
        "resetStore": [Function],
        "resetStoreCallbacks": Array [],
        "ssrMode": false,
        "store": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "cacheKeyRoot": CacheKeyNode {
              "children": null,
              "key": null,
            },
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "resultCaching": true,
            },
            "data": DepTrackingCache {
              "data": Object {},
              "depend": [Function],
            },
            "maybeBroadcastWatch": [Function],
            "optimisticData": DepTrackingCache {
              "data": Object {},
              "depend": [Function],
            },
            "silenceBroadcast": false,
            "storeReader": StoreReader {
              "cacheKeyRoot": CacheKeyNode {
                "children": null,
                "key": null,
              },
              "executeSelectionSet": [Function],
              "executeStoreQuery": [Function],
            },
            "storeWriter": StoreWriter {},
            "typenameDocumentCache": Map {},
            "watches": Set {},
          },
        },
        "version": "2.5.0-verify.4",
        "watchQuery": [Function],
      }
    }
  >
    <RegistrationForm
      errors={Array []}
      userInput={
        Object {
          "affiliation": "",
          "confirmPassword": "",
          "email": "",
          "familyName": "",
          "givenName": "",
          "password": "",
          "username": "",
        }
      }
    >
      <Mutation
        mutation={
          Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "SIGNUP_MUTATION",
                },
                "operation": "mutation",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "familyName",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "familyName",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "givenName",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "givenName",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "username",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "username",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "email",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "email",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "affiliation",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "affiliation",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "password",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "password",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "signup",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "familyName",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "givenName",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "username",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "email",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "affiliation",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "password",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 377,
              "start": 0,
            },
          }
        }
        variables={
          Object {
            "affiliation": "",
            "confirmPassword": "",
            "email": "",
            "familyName": "",
            "givenName": "",
            "password": "",
            "username": "",
          }
        }
      >
        <styled.form
          onSubmit={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [
                  Object {
                    "className": "column is-4 is-offset-4",
                  },
                ],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": true,
                  "lastClassName": "LoyGl",
                  "rules": Array [
                    "
  && {
    padding: 1px;
  }
",
                  ],
                },
                "displayName": "styled.form",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": "form",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onSubmit={[Function]}
          >
            <form
              className="sc-bdVaJa column is-4 is-offset-4 LoyGl"
              onSubmit={[Function]}
            >
              <styled.h1>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [
                        Object {
                          "className": "title is-3",
                        },
                      ],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bwzfXH",
                        "isStatic": true,
                        "lastClassName": "RJwfL",
                        "rules": Array [
                          "
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
",
                        ],
                      },
                      "displayName": "styled.h1",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bwzfXH",
                      "target": "h1",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <h1
                    className="sc-bwzfXH title is-3 RJwfL"
                  >
                    Create your account
                  </h1>
                </StyledComponent>
              </styled.h1>
              <RegistrationInputField
                label="Given name (first name)"
                name="givenName"
                onChange={[Function]}
                type="text"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="givenName+Given name (first name)"
                  >
                    Given name (first name)
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="givenName input"
                      id="givenName"
                      name="givenName"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <RegistrationInputField
                label="Family name (surname)"
                name="familyName"
                onChange={[Function]}
                type="text"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="familyName+Family name (surname)"
                  >
                    Family name (surname)
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="familyName input"
                      id="familyName"
                      name="familyName"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <RegistrationInputField
                label="Email address"
                name="email"
                onChange={[Function]}
                type="text"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="email+Email address"
                  >
                    Email address
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="email input"
                      id="email"
                      name="email"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <RegistrationInputField
                label="Username"
                name="username"
                onChange={[Function]}
                type="text"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="username+Username"
                  >
                    Username
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="username input"
                      id="username"
                      name="username"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <RegistrationInputField
                label="Affiliation"
                name="affiliation"
                onChange={[Function]}
                placeholder="E.g. University of Cape Town"
                type="text"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="affiliation+Affiliation"
                  >
                    Affiliation
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="affiliation input"
                      id="affiliation"
                      name="affiliation"
                      onChange={[Function]}
                      placeholder="E.g. University of Cape Town"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <RegistrationInputField
                label="Password"
                name="password"
                onChange={[Function]}
                placeholder="At least 7 characters"
                type="password"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="password+Password"
                  >
                    Password
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="password input"
                      id="password"
                      name="password"
                      onChange={[Function]}
                      placeholder="At least 7 characters"
                      type="password"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <RegistrationInputField
                label="Re-enter password "
                name="confirmPassword"
                onChange={[Function]}
                type="password"
                value=""
              >
                <div
                  className="field"
                >
                  <label
                    className="label"
                    htmlFor="confirmPassword+Re-enter password "
                  >
                    Re-enter password 
                    :
                  </label>
                  <div
                    className="control is-child"
                  >
                    <input
                      className="confirmPassword input"
                      id="confirmPassword"
                      name="confirmPassword"
                      onChange={[Function]}
                      type="password"
                      value=""
                    />
                  </div>
                </div>
              </RegistrationInputField>
              <button
                className="signUp button is-success is-fullwidth is-rounded"
                disabled={false}
                type="submit"
              >
                Sign up
              </button>
            </form>
          </StyledComponent>
        </styled.form>
      </Mutation>
    </RegistrationForm>
  </ApolloProvider>
</MockedProvider>
`;
